### PR TITLE
MDEV-14642 Assertion `table->s->db_create_options == part_table->s->db_create_options' failed in compare_table_with_partition

### PR DESCRIPTION
**Pull request by TakeKiyo**
- [x] *The Jira issue number for this PR is: MDEV-14642*

## Description
MDEV-14642 Assertion `table->s->db_create_options == part_table->s->db_create_options' failed in compare_table_with_partition 

When trying to execute `ALTER TABLE EXCHANGE PARTITION` with different definitions, 
assertion `table->s->db_create_options == part_table->s->db_create_options` failed in compare_table_with_partition.
However, this execution should not be allowed.

No MariaDB document about exchange partition were found, however, I found [MySQL document about exchange partition](https://dev.mysql.com/doc/refman/5.7/en/partitioning-management-exchange.html) which says 
>  Table nt is not itself partitioned.
 Table nt is not a temporary table.
 The structures of tables pt and nt are otherwise identical.

To fix the problem, I deleted the assertion code and added code that returns an error that indicates tables have different definitions. 

## How can this PR be tested?
Test case is included in this pull request. 
I add test code to `mysql/main/partition_exchange.test` since similar test code are written in this file. 

## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature and the PR is based against the latest MariaDB development branch*
- [x] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*

### CODE REVIEW

**Comment by nayuta-yanagisawa** [comment]
@TakeKiyo Thank you for the pull request. We will review it soon.

**Review by nayuta-yanagisawa — CHANGES_REQUESTED** [review]
I will check the technical details later, but I have some comments on the commit message.

* The first line of the commit message should be exactly same as the issue title.
* Please make it more descriptive (what is the problem? how to fix? etc.). More precisely, could you include what you wrote in the issue description without referencing the external document.

**Comment by TakeKiyo** [comment]
@nayuta-yanagisawa 
Thank you for your quick comment.
I modified the commit message based on your comment.

**Comment by nayuta-yanagisawa** [comment]
Pushed to bb branch: https://github.com/MariaDB/server/commit/99b90d97b0d33852117ef0c0e04ddce0f602042c

**Comment by nayuta-yanagisawa** [comment]
Just a minor comment but, in the general convention of git, you should properly insert the new lines to the body of the commit message. Please check other commits of the repository.

**Comment by TakeKiyo** [comment]
I rebased and modified the commit message.
Thank you for your comment.

**Review by nayuta-yanagisawa — DISMISSED** [review]

**Comment by nayuta-yanagisawa** [comment]
The following test case passes and thus the problem seems to be Arira/MyISAM specific:

```
--source include/have_partition.inc
--
--source include/have_innodb.inc
 
CREATE TABLE t1 (i INT NOT NULL, KEY (i)) ENGINE=InnoDB ROW_FORMAT=REDUNDANT PARTITION BY KEY(i) PARTITIONS 2;
CREATE TABLE t2 (i INT NOT NULL, KEY (i)) ENGINE=InnoDB;
--error ER_PARTITION_EXCHANGE_DIFFERENT_OPTION
ALTER TABLE t1 EXCHANGE PARTITION p1 WITH TABLE t2;
 
# Cleanup
DROP TABLE t1, t2;
```

**Comment by nayuta-yanagisawa** [comment]
I wrote down my analysis to https://jira.mariadb.org/browse/MDEV-14642. I'm now thinking of how the assertion error should be fixed. Please wait for a while.

File: mysql-test/main/partition_exchange.test [hunk 1]
```diff
@@ -536,3 +536,11 @@ ALTER TABLE t2 REMOVE PARTITIONING;
 ALTER TABLE t1 EXCHANGE PARTITION pm WITH TABLE t2;
 DROP TABLE t1, t2;
 
+--echo # Test exchange partition with different ROW_FORMAT
```
Comment (nayuta-yanagisawa) [inline mysql-test/main/partition_exchange.test:539]: Explicitly mention the issue ID and title.

```
--echo #
--echo # MDEV-14642 Assertion `table->s->db_create_options == part_table->s->db_create_options' failed in compare_table_with_partition
--echo #
```

File: mysql-test/main/partition_exchange.test [hunk 2]
```diff
@@ -536,3 +536,11 @@ ALTER TABLE t2 REMOVE PARTITIONING;
...
+CREATE TABLE t1 (a INT) ROW_FORMAT=DYNAMIC PARTITION BY KEY(a) PARTITIONS 2;
+CREATE TABLE t2 (a INT) ;
+--error ER_PARTITION_EXCHANGE_DIFFERENT_OPTION
+ALTER TABLE t1 EXCHANGE PARTITION p1 WITH TABLE t2;
+ 
+# Cleanup
+DROP TABLE t1, t2;
```
Comment (nayuta-yanagisawa) [inline mysql-test/main/partition_exchange.test:546]: Add the following test case, which is a simplified version of one provided by Roel:

```
CREATE TABLE t1 (a INT, PRIMARY KEY(a)) ENGINE=InnoDB PARTITION BY KEY(a) PARTITIONS 2;
CREATE TABLE t2 (a INT, PRIMARY KEY(a)) CHECKSUM=1, ENGINE=InnoDB;
ALTER TABLE t1 EXCHANGE PARTITION p0 WITH TABLE t2;
 
# Cleanup
DROP TABLE t1, t2;
```

**Review by nayuta-yanagisawa — CHANGES_REQUESTED** [review]
@TakeKiyo After a private discussion with @holyfoot , I now think that the fix is mostly OK while I still have some comments.

**Comment by TakeKiyo** [comment]
@nayuta-yanagisawa 
I modified the code based on your comment.

Thank you for your kind guidance.
I didn't understand the whole analysis, so I will take my time to understand it.

File: sql/sql_partition_admin.cc [hunk 3]
```diff
@@ -249,8 +249,13 @@ static bool compare_table_with_partition(THD *thd, TABLE *table,
     my_error(ER_TABLES_DIFFERENT_METADATA, MYF(0));
     DBUG_RETURN(TRUE);
   }
-  DBUG_ASSERT(table->s->db_create_options ==
-              part_table->s->db_create_options);
+
+  if (table->s->db_create_options != part_table->s->db_create_options)
```
Comment (nayuta-yanagisawa) [inline sql/sql_partition_admin.cc:253]: Unnecessary space at the end of the line

File: mysql-test/main/partition_exchange.test [hunk 4]
```diff
@@ -536,3 +536,21 @@ ALTER TABLE t2 REMOVE PARTITIONING;
...
+
+CREATE TABLE t1 (a INT, PRIMARY KEY(a)) ENGINE=InnoDB PARTITION BY KEY(a) PARTITIONS 2;
+CREATE TABLE t2 (a INT, PRIMARY KEY(a)) CHECKSUM=1, ENGINE=InnoDB;
+--error ER_TABLES_DIFFERENT_METADATA
+ALTER TABLE t1 EXCHANGE PARTITION p0 WITH TABLE t2;
+
```
Comment (nayuta-yanagisawa) [inline mysql-test/main/partition_exchange.test:554]: Unnecessary space

File: mysql-test/main/partition_exchange.test [hunk 5]
```diff
@@ -536,3 +536,21 @@ ALTER TABLE t2 REMOVE PARTITIONING;
...
 DROP TABLE t1, t2;
 
+--echo #
+--echo # MDEV-14642 Assertion `table->s->db_create_options == part_table->s->db_create_options' failed in compare_table_with_partition
+--echo #
+CREATE TABLE t1 (a INT) ROW_FORMAT=DYNAMIC PARTITION BY KEY(a) PARTITIONS 2;
+CREATE TABLE t2 (a INT) ;
+--error ER_TABLES_DIFFERENT_METADATA
+ALTER TABLE t1 EXCHANGE PARTITION p1 WITH TABLE t2;
+
```
Comment (nayuta-yanagisawa) [inline mysql-test/main/partition_exchange.test:546]: Unnecessary space

**Comment by nayuta-yanagisawa** [comment]
I pushed the commit to bb again: https://github.com/MariaDB/server/commit/9eb701ccf14e959e400a30e4c89b4c4353519879

**Comment by TakeKiyo** [comment]
@nayuta-yanagisawa 
I deleted some unnecessary spaces and pushed again. 

I couldn't notice the unnecessary space, do you use some tools to check the code?

**Comment by nayuta-yanagisawa** [comment]
@TakeKiyo I found them just by `git diff HEAD^`. `git` marks such unnecessary white spaces by red color.

**Comment by TakeKiyo** [comment]
@nayuta-yanagisawa 
Thank you.
I will check carefully before creating pr from now on.

**Pull request merged**